File: examples/ionBeamEtching/ionBeamEtching.py
from argparse import ArgumentParser
import numpy as np
import viennaps as ps

# parse config file name and simulation dimension
parser = ArgumentParser(
    prog="ionBeamEtching",
    description="Run an IBE process on a trench geometry.",
)
parser.add_argument("-D", "-DIM", dest="dim", type=int, default=2)
parser.add_argument("filename")
args = parser.parse_args()

# switch between 2D and 3D mode
if args.dim == 2:
    print("Running 2D simulation.")
else:
    print("Running 3D simulation.")
ps.setDimension(args.dim)

params = ps.readConfigFile(args.filename)

geometry = ps.Domain(
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=params["yExtent"],
)
ps.MakeTrench(
    domain=geometry,
    trenchWidth=params["trenchWidth"],
    trenchDepth=params["trenchDepth"],
    maskHeight=params["maskHeight"],
).apply()

# copy top layer to capture deposition
geometry.duplicateTopLevelSet(ps.Material.Polymer)

ibeParams = ps.IBEParameters()
ibeParams.tiltAngle = params["angle"]
ibeParams.exponent = params["exponent"]
ibeParams.thetaRMin = 0.0
ibeParams.thetaRMax = 15.0

ibeParams.meanEnergy = params["meanEnergy"]
ibeParams.sigmaEnergy = params["sigmaEnergy"]
ibeParams.thresholdEnergy = params["thresholdEnergy"]

ibeParams.redepositionRate = params["redepositionRate"]
ibeParams.planeWaferRate = params["planeWaferRate"]

model = ps.IonBeamEtching(
    parameters=ibeParams,
    maskMaterials=[ps.Material.Mask],
)

direction = [0.0, 0.0, 0.0]
direction[0] = np.sin(ibeParams.tiltAngle * np.pi / 180.0)
direction[args.dim - 1] = -np.cos(ibeParams.tiltAngle * np.pi / 180.0)
model.setPrimaryDirection(direction)

advParams = ps.AdvectionParameters()
advParams.spatialScheme = ps.SpatialScheme.LAX_FRIEDRICHS_2ND_ORDER

process = ps.Process(geometry, model)
process.setProcessDuration(params["processTime"])
process.setParameters(advParams)

geometry.saveHullMesh("initial")

process.apply()

geometry.saveHullMesh("final")

File: examples/ionBeamEtching/ionBeamEtching.cpp
#include <geometries/psMakeTrench.hpp>
#include <models/psIonBeamEtching.hpp>

#include <process/psProcess.hpp>
#include <psUtil.hpp>

using namespace viennaps;

int main(int argc, char *argv[]) {
  using NumericType = double;
  constexpr int D = 2;

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    // Try default config file
    params.readConfigFile("config.txt");
    if (params.m.empty()) {
      std::cout << "No configuration file provided!" << std::endl;
      std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
      return 1;
    }
  }

  auto geometry = Domain<NumericType, D>::New(
      params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"),
      BoundaryType::PERIODIC_BOUNDARY);
  MakeTrench<NumericType, D>(geometry, params.get("trenchWidth"),
                             params.get("trenchDepth"),
                             0.0 /*trenchTaperAngle*/, params.get("maskHeight"))
      .apply();

  // copy top layer to capture deposition
  geometry->duplicateTopLevelSet(Material::Polymer);

  IBEParameters<NumericType> ibeParams;
  ibeParams.tiltAngle = params.get("angle");
  ibeParams.exponent = params.get("exponent");
  ibeParams.thetaRMin = 0.;
  ibeParams.thetaRMax = 15.;
  ibeParams.rotatingWafer = true;

  ibeParams.meanEnergy = params.get("meanEnergy");
  ibeParams.sigmaEnergy = params.get("sigmaEnergy");
  ibeParams.thresholdEnergy = params.get("thresholdEnergy");

  ibeParams.redepositionRate = params.get("redepositionRate");
  ibeParams.planeWaferRate = params.get("planeWaferRate");

  auto model = SmartPointer<IonBeamEtching<NumericType, D>>::New(
      ibeParams, std::vector<Material>{Material::Mask});
  Vec3D<NumericType> direction{0., 0., 0.};
  direction[0] = std::sin(ibeParams.tiltAngle * M_PI / 180.);
  direction[D - 1] = -std::cos(ibeParams.tiltAngle * M_PI / 180.);
  model->setPrimaryDirection(direction);

  AdvectionParameters advectionParams;
  advectionParams.spatialScheme =
      viennals::SpatialSchemeEnum::LAX_FRIEDRICHS_2ND_ORDER;

  Process<NumericType, D> process(geometry, model);
  process.setProcessDuration(params.get("processTime"));
  process.setParameters(advectionParams);

  geometry->saveHullMesh("initial");

  process.apply();

  geometry->saveHullMesh("final");
}
